### DISCUSSION TITLE
JKBMS B2A8S30P working in parallel with Cerbo control

### DISCUSSION THREAD

**Discussion by antoniozys** [mr-manuel/venus-os_dbus-serialbattery#117] [Show and tell]
I would like to talk about my experience (a fight) to get 4 BMS of the B2A8S30P running and being controlled by the cerbo unit.
This BMS is an strange hybrid. The Serialbattery  app recognizes it as JKBMS PB but this is not the inverter version.
I suspect they share firmmware.
I can connect the JKBMS PC app and control the board through a PC , the same as the inverter version.
I have no CAN bus and I work with the RS485 serial connection through the JKBMS adaptor.
There are two models of such adaptors. With the low voltage batteriesw (4s and 8s) only works the special one requested.

**Comment by antoniozys** [comment]
When I connected the first battery through the pc everything worked as a charm inmediatelly. 
Connection with the serial battery needs to revise the config file and write down max charge and discharge and other parameters that I discovered after several days of trying.
The most important is to avoid the standard unique identifier as the serial number.
The JKBMS uses a 11 digits serial and on serial battery only 120 arrive.
Mr_Manuel show me the way: using the tty serial connection as the unique number.,
This is achieved with his sentence at the config file:
USE_PORT_AS_UNIQUE_ID = True

**Comment by antoniozys** [comment]
To accelerate the seach of the bms by the algorithm is convenient to reduce the search to the jkbms  with (config file):

BMS_TYPE = Jkbms, Jkbms_pb, Jkbms_Can

**Comment by antoniozys** [comment]
The sistem only found one battery and was frustrating till I checked to define adresses at the modbus sentence (config.ini)
MODBUS_ADDRESSES = 0x01, 0x02, 0x03, 0x04

These numbers are the one in the jkbms app that define the identity of each board as Device addr

**Reply by antoniozys** [reply]
![imagen](https://github.com/user-attachments/assets/b1a4f9ea-15b5-47f4-b5a6-15ceb3cb282a)

**Comment by antoniozys** [comment]
With these 3 lines the visibility of the batteries and the aggregate are ok finally!

**Comment by antoniozys** [comment]
![imagen](https://github.com/user-attachments/assets/5d331a30-602c-49ae-8693-453541667424)

**Reply by antoniozys** [reply]
![imagen](https://github.com/user-attachments/assets/ae1ff090-6e3f-40af-b5bb-a19d1a78a0ba)
The device unique identifier appears like this

**Comment by antoniozys** [comment]
Because I will use 4 batteries in parallel with high intensity in amperes I decided to check the paralel hardware board sold by jkms.
I'm just starting with the tests.
Looks like is possible to limit the currents between batteries and define it in the jkbms bluetooth app.
The board is connected inside each battery with some provided wires to several points of the bms
I'll check and inform of the results
IN my case: the bms can deliver 300A (max theoretical in paralel 4 units= 1200A) and and will limit to 200A each battery and total of 800A as aggregate.

**Comment by antoniozys** [comment]
Hope this can help anyone trying to connect this stange bms in parallel.

**Reply by Albarge** [reply]
Thanks for this write up @antoniozys I have the very common B2A8S20P with hardware V.11.XW so probably similar to you, what is the JK parallel hardware board? I use https://github.com/pulquero/BatteryAggregator to run 3 x parallel batteries and make them all show up and combine CCL + CVL etc with dbus-serialbattery.

So did you daisy chain your RS485 for all four batteries like this?
![daisy-chain-rs485-1-8469e98594a570bbec07609b6c6a6fcc](https://github.com/user-attachments/assets/a7fcf756-690a-4999-97ab-fb6243b566ab) Did you connect GND?

If I understand correctly, to make them work the device address and config file are slightly confusing. If we select in the BMS device 2, we need to put in the hexadecimal version in the config: BATTERY_ADDRESSES = 0x02 ?

**Comment by antoniozys** [comment]
HI
My B2A8S30P is recognised by serialbattery as jkbms PB model (same as inverter) I do not know if yours is the same. With the pc app delivered by JKBMS I can access the data in a PC through a serial usb and cable. I suspect the firmware is the same of the inverter version.
This board has an special device , the parallel board and works like this:
![JKBMS-Parallelle-Module-JK-BLMK-5A-for-JIKONG-BMS-JK-B2A8S30P-6](https://github.com/user-attachments/assets/e6007367-3665-4133-9494-835a210db652)
Theoretically compensates the load and unloads of the batteries when connected in parallel till 4 units.
![JKBMS-Parallel-Module-JK-BLMK-5A-Wiring-diagram](https://github.com/user-attachments/assets/e68f6589-6f8d-4dc0-afcf-551e4624f398)
[JK-BLMK-5A-V3.0电池并联模块使用说明书20231123_1_10_translate_20240802153717.pdf](https://github.com/user-attachments/files/18108402/JK-BLMK-5A-V3.0.20231123_1_10_translate_20240802153717.pdf)

**Comment by antoniozys** [comment]
AS far as I know this board has been designed only for the B2A8S30P that can deliver 300A ( Mines are limited by me to 200A)
Related with the daisy chain connection, this is common with the buses, CAN or RS485 or for nautical NMEA2000. 
The NMEA2000 is a can bus with 12v feed in one wire. Used in boats to connect devices and supply energy to them.
To understand how it works is better to see one piece used in boats to connect banches (2m max short)
![multiconector-nmea-2000](https://github.com/user-attachments/assets/a3133df3-ece9-4753-a4b2-e439d4a743b0)
The only important matter is to "close" the external branches of the deployement with a resistance of 120 Ohm on each side.
This avoids the reflexes of the signal at the ends and the "noise" produced. For short connections is not necessary. Better check your installation.
![imagen](https://github.com/user-attachments/assets/090a2141-deab-4746-8dc5-f2a8a7642a54)

And yes on my tests following the recommendations of Mr.Manuel I didn't ground the lines. Only using A and B wires. Ground wire not used

**Comment by antoniozys** [comment]
I have not finished yet the complete installation (now working with the internal battery wiring...) and finnally conecting the 4 units.
At least the communication with Cerbo succeed and can see all the batteries and their status.
* in my bms board I can define the unit device address. I have 1,2,3 and 4 on each battery. Check yours if is a different model or firmware.
![imagen](https://github.com/user-attachments/assets/73eeaedd-2102-4448-b075-a67ebb451758)

After using the USE_PORT_AS_UNIQUE_ID = True
The serialbattery recognizes the unit as...
![392938961-ae1ff090-6e3f-40af-b5bb-a19d1a78a0ba](https://github.com/user-attachments/assets/706ec427-3f5a-4d06-b6cf-5ffa11b27d0c)
for the unit 4
